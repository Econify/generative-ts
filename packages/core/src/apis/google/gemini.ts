/* eslint-disable no-nested-ternary */
/* eslint-disable camelcase */
import * as t from "io-ts";
import type { TypeOf } from "io-ts";
import { isLeft } from "fp-ts/lib/Either.js";

import type { ModelApi, ModelRequestOptions } from "@typeDefs";

import { FnTemplate } from "../../utils/Template";

import { composite } from "../_utils/ioTsHelpers";

import type { FewShotRequestOptions, ToolUseRequestOptions } from "../shared";

interface FunctionCall {
  name: string;
  args: Record<string, unknown>;
}

interface FunctionResponse {
  name: string;
  response: Record<string, unknown>;
}

interface Part {
  text?: string;
  functionCall?: FunctionCall; // TODO deal with casing disparity
  function_response?: FunctionResponse;
}

interface PartWithFunctionCall extends Part {
  functionCall: FunctionCall;
}

interface PartWithFunctionResponse extends Part {
  function_response: FunctionResponse;
}

interface GoogleGeminiContentItem {
  role?: "user" | "model";
  parts: Part[];
}

interface GoogleGeminiSchema {
  type: "STRING" | "INTEGER" | "BOOLEAN" | "NUMBER" | "ARRAY" | "OBJECT";
  description?: string;
  enum?: string[];
  items?: GoogleGeminiSchema[];
  properties?: {
    [key: string]: GoogleGeminiSchema;
  };
  required?: string[];
  nullable?: boolean;
}

interface GoogleGeminiToolsOptions {
  tools?: {
    function_declarations: {
      name: string;
      description?: string;
      parameters?: GoogleGeminiSchema;
      response?: GoogleGeminiSchema;
    }[];
  }[];
}

/**
 * @category Google Gemini
 * @category Requests
 */
export interface GoogleGeminiOptions
  extends ModelRequestOptions,
    FewShotRequestOptions,
    ToolUseRequestOptions,
    GoogleGeminiToolsOptions {
  contents?: GoogleGeminiContentItem | GoogleGeminiContentItem[];
  system_instruction?: GoogleGeminiContentItem;
  tools_config?: {
    mode?: "AUTO" | "NONE" | "ANY";
    allowed_function_names?: string[];
  };
  safety_settings?: {
    category?: string;
    threshold?: string;
    max_influential_terms?: number;
    method?: string;
  };
  generation_config?: {
    temperature?: number;
    top_p?: number;
    top_k?: number;
    candidate_count?: number;
    max_output_tokens?: number;
    stop_sequences?: string[];
    presence_penalty?: number;
    frequency_penalty?: number;
    response_mime_type?: "text/plain" | "application/json";
  };
}

const toGeminiToolParamType = (type: "STR" | "NUM" | "BOOL") => {
  return (
    (type === "STR" && "STRING") || (type === "NUM" && "NUMBER") || "BOOLEAN"
  );
};

function mapToolDescriptionsToGeminiRequest({
  $tools: tools,
}: ToolUseRequestOptions): GoogleGeminiToolsOptions {
  if (!tools) {
    return {
      tools: [],
    };
  }

  return {
    tools: [
      {
        function_declarations: tools.map((tool) => {
          return {
            name: tool.name,
            description: tool.description,
            ...(tool.parameters
              ? {
                  parameters: {
                    type: "OBJECT",
                    properties: tool.parameters.reduce(
                      (acc, param) => {
                        acc[param.name] = {
                          type: toGeminiToolParamType(param.type),
                          description: param.description,
                        };
                        return acc;
                      },
                      {} as { [key: string]: GoogleGeminiSchema },
                    ),
                    required: tool.parameters
                      .filter(({ required }) => required)
                      .map(({ name }) => name),
                  },
                }
              : {}),
          };
        }),
      },
    ],
  };
}

/**
 * @category Google Gemini
 * @category Templates
 */
export const GoogleGeminiTemplate = new FnTemplate(
  ({
    prompt,
    examplePairs,
    system,
    $tools,
    contents,
    tools,
    tools_config,
    system_instruction,
    safety_settings,
    generation_config,
  }: GoogleGeminiOptions) => {
    const _contents: GoogleGeminiContentItem[] = [
      ...(examplePairs
        ? examplePairs.flatMap((pair) => [
            {
              role: "user" as const,
              parts: [{ text: pair.user }],
            },
            {
              role: "model" as const,
              parts: [{ text: pair.assistant }],
            },
          ])
        : []),

      ...(contents ? (Array.isArray(contents) ? contents : [contents]) : []),
    ];

    const lastItem = _contents[_contents.length - 1];

    if (!lastItem) {
      _contents.push({
        role: "user",
        parts: [{ text: prompt }],
      });
    } else if (lastItem.role === "model") {
      const functionCalls = lastItem.parts
        .filter((part): part is PartWithFunctionCall => "functionCall" in part)
        .map((part) => part.functionCall);

      if (functionCalls.length && $tools) {
        const responses: PartWithFunctionResponse[] = [];

        functionCalls.forEach(({ name, args }) => {
          const matchingTool = $tools.find((tool) => tool.name === name);

          if (!matchingTool) {
            console.warn(
              "The last item of conversation history (`contents`) contains a `function_call`, but no matching tool was found in `$tools`. Model behavior might be unexpected, because model's function calls were effectively ignored.",
            );

            return;
          }

          const matchingInvocation = matchingTool.invocations
            ?.reverse()
            .find((invocation) =>
              Object.keys(args).every(
                (key) =>
                  key in invocation.arguments &&
                  invocation.arguments[key] === args[key],
              ),
            );

          if (!matchingInvocation) {
            console.warn(
              "The last item of conversation history (`contents`) contains a `function_call`, and a matching `$tool` was found, but no matching invocation was found in the tool's invocations. (Did you forget to call `mapGeminiResponseToToolInvocations`?) Model behavior might be unexpected, because model's function calls were effectively ignored.",
            );

            return;
          }

          const { returned } = matchingInvocation;

          if (!returned) {
            console.warn(
              "The last item of conversation history (`contents`) contains a `function_call`, and a $tool with matching invocations was found, but that invocation does NOT have a `returned` value! (Did you forget to execute the tool?) Model behavior might be unexpected, because model's function calls were effectively ignored.",
            );

            return;
          }

          responses.push({
            function_response: {
              name: matchingTool.name,
              response: {
                returned,
              },
            },
          });
        });

        _contents.push({
          role: "user",
          parts: responses,
        });
      } else {
        if (functionCalls.length && !$tools) {
          console.warn(
            "The last item of conversation history (`contents`) contains a `function_call`, but no `$tools` were passed, so generative-ts cannot append `function_response` to conversation history. Instead, appending prompt to end of conversation history. Model behavior might be unexpected, because model's function calls were effectively ignored.",
          );
        }

        _contents.push({
          role: "user",
          parts: [{ text: prompt }],
        });
      }
    }

    const rewritten = {
      contents: _contents,
      ...(tools || $tools
        ? {
            tools: [
              ...(tools || []),
              ...($tools
                ? mapToolDescriptionsToGeminiRequest({ $tools }).tools || []
                : []),
            ],
          }
        : {}),
      ...(system_instruction || system
        ? {
            system_instruction: {
              parts: [
                ...(system ? [{ text: system }] : []),
                ...(system_instruction
                  ? system_instruction.parts.map((part) => ({
                      ...(part.text ? { text: part.text } : {}),
                    }))
                  : []),
              ],
            },
          }
        : {}),
      ...(tools_config
        ? {
            tools_config,
          }
        : {}),
      ...(safety_settings
        ? {
            safety_settings,
          }
        : {}),
      ...(generation_config
        ? {
            generation_config,
          }
        : {}),
    };

    return JSON.stringify(rewritten, null, 2);
  },
);

const GoogleGeminiResponseCodec = t.type({
  data: t.type({
    candidates: t.array(
      composite({
        required: {
          finishReason: t.string,
        },
        partial: {
          content: t.type({
            role: t.string,
            parts: t.array(
              t.partial({
                text: t.string,
                functionCall: t.type({
                  name: t.string,
                  args: t.record(t.string, t.unknown),
                }),
              }),
            ),
          }),
          citationMetadata: t.type({
            citations: t.array(
              t.partial({
                uri: t.string,
                startIndex: t.number,
                endIndex: t.number,
              }),
            ),
          }),
          safetyRatings: t.array(
            t.type({
              category: t.string,
              probability: t.string,
              probabilityScore: t.number,
              severity: t.string,
              severityScore: t.number,
            }),
          ),
        },
      }),
    ),
    usageMetadata: t.type({
      candidatesTokenCount: t.number,
      promptTokenCount: t.number,
      totalTokenCount: t.number,
    }),
  }),
  headers: t.record(t.string, t.unknown),
  status: t.number,
  statusText: t.string,
  // config
  // request
});

/**
 * @category Google Gemini
 * @category Responses
 */
export interface GoogleGeminiResponse
  extends TypeOf<typeof GoogleGeminiResponseCodec> {}

export function isGoogleGeminiResponse(
  response: unknown,
): response is GoogleGeminiResponse {
  return !isLeft(GoogleGeminiResponseCodec.decode(response));
}

export interface GoogleGeminiApi
  extends ModelApi<GoogleGeminiOptions, GoogleGeminiResponse> {}

/**
 *
 * ## Reference
 * - {@link https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/inference | Gemini Inference API}
 * - {@link https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/function-calling | Gemini Function Calling API}
 *
 * ## Providers using this API
 * - {@link createVertexAiModelProvider | GCloud VertexAI}
 *
 * @category APIs
 * @category Google Gemini
 * @category Provider: GCloud VertexAI
 *
 */
export const GoogleGeminiApi: GoogleGeminiApi = {
  requestTemplate: GoogleGeminiTemplate,
  responseGuard: isGoogleGeminiResponse,
};
