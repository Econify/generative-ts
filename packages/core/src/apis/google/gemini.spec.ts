import { GoogleGeminiApi, GoogleGeminiOptions } from "./gemini";

function render(context: unknown) {
  try {
    const rendered = GoogleGeminiApi.requestTemplate.render(
      context as GoogleGeminiOptions,
    );
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("GoogleGeminiApi.requestTemplate", () => {
  test("prompt", () => {
    const rendered = render({
      prompt: "mock-prompt",
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
    });
  });

  test("prompt, system", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system: "mock-system-text",
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
      system_instruction: {
        parts: [
          {
            text: "mock-system-text",
          },
        ],
      },
    });
  });

  test("prompt, system_instruction", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system_instruction: {
        parts: [{ text: "mock-system-text" }],
      },
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
      system_instruction: {
        parts: [{ text: "mock-system-text" }],
      },
    });
  });

  test("prompt, system and system_instruction", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system: "mock-system-text",
      system_instruction: {
        parts: [
          { text: "mock-additional-instruction" },
          { text: "mock-additional-instruction-2" },
        ],
      },
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
      system_instruction: {
        parts: [
          {
            text: "mock-system-text",
          },
          {
            text: "mock-additional-instruction",
          },
          {
            text: "mock-additional-instruction-2",
          },
        ],
      },
    });
  });

  test("prompt, examplePairs", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-user-msg-1" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-assistant-msg-1" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-user-msg-2" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-assistant-msg-2" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
    });
  });

  test("prompt, contents", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-model-text-2" }],
        },
      ],
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-model-text-2" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
    });
  });

  test("prompt, examplePairs, contents", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
      ],
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-user-msg-1" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-assistant-msg-1" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
    });
  });

  test("doesnt include prompt if contents end with a user role", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
      ],
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
      ],
    });
  });

  test("prompt, contents with function_call", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          parts: [
            {
              function_call: {
                name: "mock-function",
                args: { key: "value" },
              },
            },
          ],
        },
      ],
    });

    expect(rendered).toEqual({
      contents: [
        {
          parts: [
            {
              function_call: {
                name: "mock-function",
                args: { key: "value" },
              },
            },
          ],
        },
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
    });
  });

  test("prompt, contents with function_response", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          parts: [
            {
              function_response: {
                name: "mock-function",
                response: { key: "value" },
              },
            },
          ],
        },
      ],
    });

    expect(rendered).toEqual({
      contents: [
        {
          parts: [
            {
              function_response: {
                name: "mock-function",
                response: { key: "value" },
              },
            },
          ],
        },
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
    });
  });

  test("prompt, tools", () => {
    const rendered = render({
      prompt: "mock-prompt",
      tools: [
        {
          function_declarations: [
            {
              name: "mock-function",
              description: "mock-description",
              parameters: {
                type: "OBJECT",
                properties: {
                  key: { type: "STRING" },
                },
              },
            },
          ],
        },
      ],
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
      tools: [
        {
          function_declarations: [
            {
              name: "mock-function",
              description: "mock-description",
              parameters: {
                type: "OBJECT",
                properties: {
                  key: { type: "STRING" },
                },
              },
            },
          ],
        },
      ],
    });
  });

  test("prompt, tools_config", () => {
    const rendered = render({
      prompt: "mock-prompt",
      tools_config: {
        mode: "AUTO",
        allowed_function_names: ["mock-function"],
      },
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
      tools_config: {
        mode: "AUTO",
        allowed_function_names: ["mock-function"],
      },
    });
  });

  test("prompt, safety_settings", () => {
    const rendered = render({
      prompt: "mock-prompt",
      safety_settings: {
        category: "mock-category",
        threshold: "mock-threshold",
      },
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
      safety_settings: {
        category: "mock-category",
        threshold: "mock-threshold",
      },
    });
  });

  test("prompt, generation_config", () => {
    const rendered = render({
      prompt: "mock-prompt",
      generation_config: {
        temperature: 0.7,
        top_p: 0.9,
      },
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
      generation_config: {
        temperature: 0.7,
        top_p: 0.9,
      },
    });
  });

  test("prompt, tool, tools_config, system_instruction, safety_settings, generation_config", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
      ],
      tools: [
        {
          function_declarations: [
            {
              name: "mock-function",
            },
          ],
        },
      ],
      tools_config: {
        mode: "ANY",
      },
      system_instruction: {
        parts: [{ text: "mock-system-text" }],
      },
      safety_settings: {
        threshold: "mock-threshold",
        method: "mock-method",
      },
      generation_config: {
        top_p: 0.7,
        max_output_tokens: 100,
      },
    });

    expect(rendered).toEqual({
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-prompt" }],
        },
      ],
      tools: [
        {
          function_declarations: [
            {
              name: "mock-function",
            },
          ],
        },
      ],
      tools_config: {
        mode: "ANY",
      },
      system_instruction: {
        parts: [{ text: "mock-system-text" }],
      },
      safety_settings: {
        threshold: "mock-threshold",
        method: "mock-method",
      },
      generation_config: {
        top_p: 0.7,
        max_output_tokens: 100,
      },
    });
  });
});
