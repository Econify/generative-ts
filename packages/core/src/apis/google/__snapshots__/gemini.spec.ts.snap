// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleGeminiApi.requestTemplate prompt 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, $tools 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "mock-description-1",
          "name": "mock-function-1",
          "parameters": {
            "properties": {
              "mock-function-1-param-1": {
                "description": "mock-function-1-param-1-description-1",
                "type": "STRING",
              },
              "mock-function-1-param-2": {
                "description": "mock-function-1-param-2-description-2",
                "type": "NUMBER",
              },
            },
            "required": [],
            "type": "OBJECT",
          },
        },
        {
          "description": "mock-description-2",
          "name": "mock-function-2",
          "parameters": {
            "properties": {
              "mock-function-2-param-1": {
                "description": "mock-function-2-param-1-description-1",
                "type": "BOOLEAN",
              },
            },
            "required": [],
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, contents (appends prompt) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text-2",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, contents ending with function_call, $tools with matching invocation (adds function_response content items) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "function_call": {
            "args": {
              "key": "value",
            },
            "name": "mock-function",
          },
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "function_response": {
            "name": "mock-function",
            "response": {
              "returned": {
                "responseKey": "responseValue",
              },
            },
          },
        },
      ],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "mock-description",
          "name": "mock-function",
          "parameters": {
            "properties": {
              "key": {
                "description": "mock-key-description",
                "type": "STRING",
              },
            },
            "required": [
              "key",
            ],
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, contents ending with model function_call, $tools without matching invocation (appends prompt; TODO logs warning) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "function_call": {
            "args": {
              "key": "value",
            },
            "name": "mock-function",
          },
        },
      ],
      "role": "model",
    },
    {
      "parts": [],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "another-description",
          "name": "another-function",
          "parameters": {
            "properties": {
              "another-key": {
                "description": "another-key-description",
                "type": "STRING",
              },
            },
            "required": [],
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, contents ending with model function_call, no $tools (appends prompt; TODO logs warning) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "function_call": {
            "args": {
              "key": "value",
            },
            "name": "mock-function",
          },
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, contents ending with user (does not append prompt) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text-2",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text-2",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, contents ending with user function_response (doesnt append prompt) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "function_call": {
            "args": {
              "key": "value",
            },
            "name": "mock-function",
          },
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "function_response": {
            "name": "mock-function",
            "response": {
              "responseKey": "responseValue",
            },
          },
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, contents, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text-2",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, examplePairs 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-msg-1",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-1",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-msg-2",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-2",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, examplePairs, contents 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-msg-1",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-1",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, examplePairs, contents, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-msg-1",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-1",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, examplePairs, system 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-msg-1",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-1",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, examplePairs, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-msg-1",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-1",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, generation_config 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "generation_config": {
    "temperature": 0.7,
    "top_p": 0.9,
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, safety_settings 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "safety_settings": {
    "category": "mock-category",
    "threshold": "mock-threshold",
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, system 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, system, contents 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, system, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
      {
        "text": "mock-additional-instruction",
      },
      {
        "text": "mock-additional-instruction-2",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, tool, tools_config, system_instruction, safety_settings, generation_config 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "generation_config": {
    "max_output_tokens": 100,
    "top_p": 0.7,
  },
  "safety_settings": {
    "method": "mock-method",
    "threshold": "mock-threshold",
  },
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
  "tools": [
    {
      "function_declarations": [
        {
          "name": "mock-function",
        },
      ],
    },
  ],
  "tools_config": {
    "mode": "ANY",
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, tools 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "mock-description",
          "name": "mock-function",
          "parameters": {
            "properties": {
              "key": {
                "type": "STRING",
              },
            },
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, tools, $tools 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "mock-description",
          "name": "mock-function",
          "parameters": {
            "properties": {
              "key": {
                "type": "STRING",
              },
            },
            "type": "OBJECT",
          },
        },
      ],
    },
    {
      "function_declarations": [
        {
          "description": "mock-description-1",
          "name": "mock-function-1",
          "parameters": {
            "properties": {
              "mock-function-1-param-1": {
                "description": "mock-function-1-param-1-description-1",
                "type": "STRING",
              },
              "mock-function-1-param-2": {
                "description": "mock-function-1-param-2-description-2",
                "type": "NUMBER",
              },
            },
            "required": [
              "mock-function-1-param-2",
            ],
            "type": "OBJECT",
          },
        },
        {
          "description": "mock-description-2",
          "name": "mock-function-2",
          "parameters": {
            "properties": {
              "mock-function-2-param-1": {
                "description": "mock-function-2-param-1-description-1",
                "type": "BOOLEAN",
              },
            },
            "required": [],
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate prompt, tools_config 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "tools_config": {
    "allowed_function_names": [
      "mock-function",
    ],
    "mode": "AUTO",
  },
}
`;
